trie/trie.go

```diff
@@ -899,34 +935,74 @@ func (t *Trie) tryRevive(n node, key []byte, nub MPTProofNub) (node, bool, error
 				return nub.n1, true, nil
 			}
 		}
-
 		return nil, false, nil
+	} else if len(key) == 0 && !isExpired {
+		return nil, false, fmt.Errorf("key %v not found", key)
+	} else if len(key) != 0 && isExpired {
+		return nil, false, &ExpiredNodeError{
+			ExpiredNode: n,
+			Path:        path,
+			Epoch:       0, // Set default value, will change later
+		}
 	}
+
+	// if len(key) == 0 && isExpired {
+	// 	if hashNode, ok := n.(hashNode); ok {
+	// 		cachedHash, _ := nub.n1.cache()
+	// 		if bytes.Equal(cachedHash, hashNode) {
+	// 			if n1, ok := nub.n1.(*fullNode); ok {
+	// 				n1.setEpoch(t.currentEpoch)
+	// 			}
+	// 			if nub.n2 != nil {
+	// 				switch n1 := nub.n1.(type) {
+	// 				case *shortNode:
+	// 					n1.Val = nub.n2
+	// 				default:
+	// 					return nil, false, fmt.Errorf("invalid node type")
+	// 				}
+	// 			}
+
+	// 			return nub.n1, true, nil
+	// 		}
+	// 	}
+
+	// 	return nil, false, nil
+	// }
+
 	switch n := n.(type) {
 	case *shortNode:
 		if len(key) < len(n.Key) || !bytes.Equal(key[:len(n.Key)], n.Key) {
 			return nil, false, fmt.Errorf("key %v not found", key)
 		}
-		newNode, didResolve, err := t.tryRevive(n.Val, key[len(n.Key):], nub)
+		newNode, didResolve, err := t.tryRevive(n.Val, key[len(n.Key):], nub, append(path, key[:len(n.Key)]...), isExpired)
 		if didResolve && err == nil {
 			n = n.copy()
 			n.Val = newNode
 		}
 		return n, didResolve, err
 	case *fullNode:
 		childIndex := int(key[0])
-		newNode, didResolve, err := t.tryRevive(n.Children[childIndex], key[1:], nub)
+		// TODO (asyukii): Check if child has expired
+		isExpired, _ := n.ChildExpired(nil, childIndex, t.currentEpoch) // TODO (asyukii): t.currentEpoch or t.root.getEpoch()?
+		newNode, didResolve, err := t.tryRevive(n.Children[childIndex], key[1:], nub, append(path, key[0]), isExpired)
 		if didResolve && err == nil {
 			n = n.copy()
 			n.Children[childIndex] = newNode
+			n.UpdateChildEpoch(childIndex, t.currentEpoch)
+		}
+
+		if e, ok := err.(*ExpiredNodeError); ok {
+			e.Epoch = n.GetChildEpoch(childIndex)
+			return n, didResolve, e
 		}
+
 		return n, didResolve, err
 	case hashNode:
-		tn, err := t.resolveHash(n, nil) // TODO (asyukii): Revisit epoch index
+		tn, err := t.resolveHash(n, nil)
 		if err != nil {
 			return nil, false, err
 		}
-		return t.tryRevive(tn, key, nub)
+		return t.tryRevive(tn, key, nub, path, isExpired)
```

Comment:
May miss didResolve? In revive, didResolve mean resolveHash or reviveTrieNode?

Reply:
in this context, didResolve means reviveTrieNode.

changed "didResolve" to "didRevive" for better understanding

Reply:
but when it resolves HashNode, there needs copy too?

Reply:
it will only copy the resolved hash node if the nub is revived successfully